src/seals/base_envs.py

```diff
@@ -62,19 +55,20 @@ def reward(self, state: State, action: Action, new_state: State) -> float:
     def terminal(self, state: State, step: int) -> bool:
         """Is the state terminal?"""
 
+    @abc.abstractmethod
     def obs_from_state(self, state: State) -> Observation:
         """Sample observation for given state."""
 
+    @property
+    @abc.abstractmethod
+    def observation_space(self) -> gym.Space:
```

Comment:
Why is this an abstract method but `state_space` and `action_space` concrete methods reading from instance variables?

I can see pros & cons of both designs, but we should be consistent.

Reply:
Ah I think I understand, you need to do this to make it a mix-in? I'm still a bit bothered by the inconsistency though. Maybe we should make state and action space abstract then.

Reply:
I don't find it inconsistent; we expect every Environment to have a concrete state_space and action_space that has to be provided; this is not the same for observation_space (which in most cases will be equal to the state_space)
Personally, I do think it would make sense to have a default `observation_space` and default `obs_from_state`, but we cannot do that with the typing, because we don't know what Observation is during the definition.

I also think it is weird to have `obs_from_state` as abstract but not `observation_space`; IMO obs_from_state and observation_space should be defined at the same place.

Reply:
Actually, I don't have that strong views on this, I think I'm fine with always having a self._observation_space (but then we should make it non-optional in the constructor, and keep the ResettableMDP instead of the FullyObservableMixin)
I think in that case I might also prefer just having actual attributes instead of property methods (i.e. self.action_space = action_space in __init__)

Reply:
I intended `observation_space` to be non-optional in the constructor; the current docstring and `= None` are a relic from a previous iteration I forgot to remove.

Properties have the benefit of users not being able to overwrite them by accident.